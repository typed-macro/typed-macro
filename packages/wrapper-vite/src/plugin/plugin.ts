import { Plugin } from 'vite'
import { MacroProvider } from '@typed-macro/core'
import {
  createRuntime,
  FilterOptions,
  Runtime,
  TransformerOptions,
} from '@typed-macro/runtime'
import { createEnvContext } from './impl/env'
import { InternalModules } from './impl/modules'
import { join } from 'path'
import { WatchOptions } from 'chokidar'

export type MacroPlugin = Plugin & {
  /**
   * Register macro providers to this macro manager so that
   * all macros in providers and plugins share the same runtime.
   *
   * For macro plugins:
   *  > Some options like `maxTraversals` or `typesPath` will be overridden by
   *  > manager's, `parserPlugins` will be merged with the manager's one.
   *  >
   *  > After registered, the original macro plugin will be attached to the manager,
   *  > which means no need to add the plugin to Vite/Rollup 's plugins array again.
   * @param sources macro providers or plugins.
   */
  use(...sources: MacroProvider[]): MacroPlugin
}

export type MacroPluginOptions = FilterOptions &
  TransformerOptions & {
    /**
     * The path of the automatically generated type declaration file.
     *
     * @default '<projectDir>/macros.d.ts'
     */
    typesPath?: string

    /**
     * Is in dev mode.
     *
     * @default mode !== 'production'
     * @see https://vitejs.dev/guide/env-and-mode.html#modes
     */
    dev?: boolean

    /**
     * Is in SSR mode.
     *
     * @default whether there is an SSR configuration
     * @see https://vitejs.dev/guide/ssr.html
     */
    ssr?: boolean

    /**
     * Configure chokidar FSWatcher.
     *
     * @see https://github.com/paulmillr/chokidar#api
     */
    watcherOptions?: WatchOptions
  
    /**
     * Adjust the application order.
     *
     * @see https://vitejs.dev/guide/api-plugin.html#plugin-ordering
     * @default 'pre'
     */
    enforce?: 'pre' | 'post' | undefined
  }

/**
 * Create macro plugin.
 *
 * For example,
 * ```typescript
 * // vite.config.ts
 *
 * export default defineConfig({
 *   plugins: [
 *     createMacroPlugin({ ... })
 *       .use(provideSomeMacros({ ... }))
 *   ],
 * })
 * ```
 */
export function createMacroPlugin(
  /* istanbul ignore next */
  options: MacroPluginOptions = {}
): MacroPlugin {
  const {
    exclude,
    include,
    maxTraversals,
    typesPath,
    dev,
    ssr,
    watcherOptions,
    parserPlugins,
    enforce,
  } = options

  const uninstantiatedProviders: MacroProvider[] = []

  let runtime: Runtime | undefined

  const plugin: MacroPlugin = {
    use(...sources) {
      uninstantiatedProviders.push(...sources)
      return plugin
    },
    name: 'vite-plugin-macro',
    enforce: options.hasOwnProperty('enforce') ? enforce : 'pre',
    configResolved: async (config) => {
      // create env
      const env = createEnvContext(
        dev ?? !config.isProduction,
        ssr ?? !!config.build.ssr,
        watcherOptions
      )

      // init runtime
      runtime = createRuntime(env, {
        filter: { exclude, include },
        transformer: { maxTraversals, parserPlugins },
      })

      // add providers
      uninstantiatedProviders.forEach((provider) => {
        runtime!.appendProvider(provider)
      })

      // call onStart hook and render types
      await Promise.all([
        runtime!.start(),
        runtime!.renderTypes(
          /* istanbul ignore next */
          typesPath || join(env.projectPath[0], 'macros.d.ts')
        ),
      ])
    },
    configureServer: async (server) => {
      ;(runtime!.internal.env.modules as InternalModules).__setServer(server)
    },
    resolveId: (id) => runtime?.resolveId(id),
    load: (id) => runtime?.load(id),
    transform: async (code, id) => {
      if (!(await runtime?.filter(id))) return
      const result = await runtime?.transform(code, id)
      return result && { code: result, map: null }
    },
    buildEnd: async () => {
      await runtime!.stop()
    },
  }

  return plugin
}
