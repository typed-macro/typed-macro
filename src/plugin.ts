import type { Plugin, ViteDevServer } from 'vite'
import { mkdir, readFile, writeFile } from 'fs/promises'
import { dirname, resolve } from 'path'
import { render } from 'mustache'
import { getProcessor, NamespacedMacros } from '@/processor'
import { MacroMeta } from '@/macro'
import { ParserPlugin } from '@babel/parser'
import { DevServerHelper, getDevServerHelper } from '@/helper'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type CustomTypes = {
  [namespace: string]: {
    moduleScopeTypes?: string
    macroScope: {
      name: string
      meta: MacroMeta
    }[]
  }
}

type NamespacedModules = {
  [namespace: string]: string
}

export type InternalPluginOptions = {
  name: string
  dtsPath: string

  macros: NamespacedMacros
  modules: NamespacedModules

  maxRecursion: number
  parserPlugins: ParserPlugin[]

  customTypes: CustomTypes

  hooks: MacroPluginHooks
}

export function plugin(options: InternalPluginOptions): Plugin {
  const {
    name,
    dtsPath,
    modules,
    macros,
    maxRecursion,
    parserPlugins,
    customTypes,
    hooks: {
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  const macroNamespaces = Object.keys(macros)
  const moduleNamespaces = Object.keys(modules)

  let devMode = false

  const process = getProcessor({
    parserPlugins,
    macros,
    maxRecursion,
  })

  return {
    name,
    enforce: 'pre',
    configResolved(config) {
      if (config.env.DEV) devMode = true
      generateDts(dtsPath, customTypes).then()
      // hook
      return configResolved?.(config)
    },
    resolveId(id, importer, options, ssr) {
      if (macroNamespaces.includes(id)) return id
      if (moduleNamespaces.includes(id)) return id
      // hook
      return resolveId?.bind(this)(id, importer, options, ssr)
    },
    load(id, ssr) {
      if (macroNamespaces.includes(id)) return 'export {}'
      if (moduleNamespaces.includes(id)) return modules[id]
      // hook
      return load?.bind(this)(id, ssr)
    },
    transform(code, id, ssr) {
      if (/\.[jt]sx?$/.test(id)) {
        const transformed = process(code, id, devMode)
        if (transformed !== undefined)
          return transform
            ? transform.bind(this)(transformed, id, ssr)
            : transformed
      }
      // hook
      return transform?.bind(this)(code, id, ssr)
    },
    configureServer(server) {
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    ...otherHooks,
  }
}

async function generateDts(targetPath: string, customTypes: CustomTypes) {
  await mkdir(dirname(targetPath), { recursive: true })
  await writeFile(targetPath, await renderTypes(customTypes))
}

export async function renderTypes(customTypes: CustomTypes) {
  const namespaces = Object.keys(customTypes)
  return render(
    (await readFile(resolve(__dirname, './client.d.ts.tpl'))).toString(),
    {
      namespaces: namespaces.map((ns) => ({
        module: ns,
        macros: customTypes[ns].macroScope.map((m) => ({
          macroScopeTypes: m.meta.types,
          signature: m.meta.signatures,
          name: m.name,
        })),
        moduleScopeTypes: customTypes[ns].moduleScopeTypes,
      })),
    }
  )
}
