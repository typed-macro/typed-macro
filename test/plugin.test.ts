import { CustomTypes, plugin, renderTypes } from '@/plugin'
import { Plugin } from 'vite'

describe('renderTypes()', () => {
  it('should work', (done) => {
    async function test() {
      const types: CustomTypes = {
        '@test': {
          moduleScopeTypes: 'type B = string',
          macroScope: [
            {
              name: 'test',
              meta: {
                types: ['type A = string'],
                signatures: [
                  {
                    signature: '(): void',
                    comment: 'just test',
                  },
                ],
              },
            },
          ],
        },
      }
      expect(await renderTypes(types)).toMatchSnapshot()
    }
    test().then(done)
  })
})

describe('plugin()', () => {
  let _plugin: Plugin
  beforeEach(() => {
    _plugin = plugin({
      name: 'test',
      dtsPath: '',
      modules: {
        '@helper': `export function hello(msg) {console.log(msg)}`,
      },
      macros: {
        '@echo': [
          {
            name: 'echo',
            apply: ({ path }, { template }, { prependToBody }) => {
              prependToBody(
                template.statements.ast(
                  `import { hello } from '@helper'; hello('world')`
                )
              )
              path.remove()
            },
          },
        ],
      },
      maxRecursion: 5,
      parserPlugins: [],
      customTypes: {},
      hooks: {},
    })
  })
  it('should transform with macro', () => {
    expect(
      _plugin.transform!.call(null as any, `import { echo } from '@echo'`, '')
    ).toBeUndefined()
    expect(
      _plugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toMatchSnapshot()
  })
  it('should load properly', () => {
    expect(_plugin.load!.call(null as any, '@helper')).toMatchSnapshot()
    expect(_plugin.load!.call(null as any, '@echo')).toMatchSnapshot()
  })
  it('should resolve id properly', () => {
    expect(_plugin.resolveId!.call(null as any, '@helper', '', {})).toBe(
      '@helper'
    )
    expect(_plugin.resolveId!.call(null as any, '@echo', '', {})).toBe('@echo')
  })
})
