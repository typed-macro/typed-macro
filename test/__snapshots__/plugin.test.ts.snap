// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin() should load properly 1`] = `"export function hello(msg) {console.log(msg)}"`;

exports[`plugin() should load properly 2`] = `"export {}"`;

exports[`plugin() should transform with macro 1`] = `"import { hello } from '@helper';hello('world');"`;

exports[`renderTypes() should work 1`] = `
"declare module '@test' {
  type B = string
  type A = string
  /** just test **/
  export function test(): void
}
"
`;
